kernel/src/log_segment.rs

```diff
@@ -218,14 +224,157 @@ impl LogSegment {
+    /// sidecar files contain the actual add/remove actions that would otherwise be
+    /// stored directly in the checkpoint. The sidecar file batches replace the checkpoint
+    /// batch in the top level iterator to be returned.
+    fn create_checkpoint_stream(
+        engine: &dyn Engine,
+        checkpoint_read_schema: SchemaRef,
+        meta_predicate: Option<ExpressionRef>,
+        checkpoint_parts: Vec<FileMeta>,
+        log_root: Url,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
+        // We checked that checkpoint_parts is not empty before calling this function
+        require!(
+            !checkpoint_parts.is_empty(),
+            Error::generic("Checkpoint parts should not be empty")
+        );
+
+        let need_file_actions = checkpoint_read_schema.contains(ADD_NAME);
+        require!(
+            !need_file_actions || checkpoint_read_schema.contains(SIDECAR_NAME),
+            Error::generic(
+                "If the checkpoint read schema contains file actions, it must contain the sidecar column"
+            )
+        );
+
+        let is_json_checkpoint = checkpoint_parts[0].location.path().ends_with(".json");
+        let actions = Self::read_checkpoint_files(
+            engine,
+            checkpoint_parts.clone(),
+            checkpoint_read_schema.clone(),
+            meta_predicate,
+            is_json_checkpoint,
+        )?;
+
+        // 1. In the case where the schema does not contain add/remove actions, we return the checkpoint
+        // batch directly as sidecar files only have to be read when the schema contains add/remove actions.
+        // 2. Multi-part checkpoint batches never have sidecar actions, so the batch is returned as-is.
+        if !need_file_actions || checkpoint_parts.len() > 1 {
+            return Ok(Left(actions.map_ok(|batch| (batch, false))));
+        }
+
+        let parquet_handler = engine.get_parquet_handler().clone();
+
+        // Process checkpoint batches with potential sidecar file references that need to be read
+        // to extract add/remove actions from the sidecar files.
+        let actions_iter = actions
+            // Flatten the new batches returned. The new batches could be:
+            // - the checkpoint batch itself if no sidecar actions are present in the batch
+            // - one or more sidecar batches read from the sidecar files if sidecar actions are present
+            // After flattenning, perform a map operation to return the actions and a boolean flag indicating
+            // that the batch was not read from a log file.
+            .map(move |batch_result| {
+                let checkpoint_batch = batch_result?;
+
+                Self::process_single_checkpoint_batch(
+                    parquet_handler.clone(),
+                    log_root.clone(),
+                    checkpoint_batch,
+                )
```

Comment:
nit: no need for the extra variable?
```suggestion
            .map(move |checkpoint_batch| {
                Self::process_single_checkpoint_batch(
                    parquet_handler.clone(),
                    log_root.clone(),
                    checkpoint_batch?,
                )
```